### DISCUSSION TITLE
PlayPlay endpoint reversing - Working Discussion

### DISCUSSION THREAD

**Discussion by cth-latest** [librespot-org/librespot-java#421] [General]
Hi everyone, I'm opening this thread for those who would like to participate in the reversing of the new Spotify endpoint playplay. 

The points to clarify are as follows:

- [x] What is the sent token ?
- @SuisChan: "Token is hardcoded, and not hard to find. Each version, even architecture, has its own uniq token"
- [ ] What does the key received correspond to?
- [ ] How is it obfuscated? (How to deobfuscate mainly)
- [ ] What is the cache_id and is it mandatory?
- The cache_id field is not mandotary, we still need to confirm what is it. (Prob. downloaded songs or cached ones)

All these data are sent in the playplay proto (See the playplay branch).

NOTE: No of these data is mandatory ! When sniffing HTTP request from client, i noticed that some songs only send GET request so no data at all.

Actually, I could find several hash and obfuscation functions in the decompiled sources of the mobile application but without finding a direct link with playplay.

Any help will be welcome. Thanks to all who will participate.

**Comment by SuisChan** [comment]
CacheID? Or do you mean token, the requests I've seen only have a token, but not a CacheID. Can you give an example ?

**Reply by cth-latest** [reply]
As you can see on the proto added by @devgianlu on the playplay branch (playplay.proto), here is a cache_id field.

```proto
message PlayPlayLicenseRequest {
    optional int32 version = 1;
    optional bytes token = 2;
    optional bytes cache_id = 3;
    optional Interactivity interactivity = 4;
    optional ContentType content_type = 5;
    optional int64 timestamp = 6;
}
```

We now know that it's not mandotary at all, as i can confirm it was not sent by the client on my test. 

Protobuf Inspector Output:
```shell
    1 <varint> = 2
    2 <chunk> = bytes (16)
        0000   01 51 CF 56 53 80 0E AF 9A CB 39 EA AE C7 6F 15                          .Q.VS.....9...o.
    4 <varint> = 1
    5 <varint> = 1
    6 <varint> = 1633260632
```

It's probably refering to downloaded songs or cached ones. (EDIT: Not sure as when testing with recently listened songs, no data was sent at all)

**Reply by cth-latest** [reply]
@SuisChan I have added the token as a necessary search

**Reply by SuisChan** [reply]
> @SuisChan I have added the token as a necessary search

Token is hardcoded, and not hard to find. Each version, even architecture, has its own uniq token

**Reply by cth-latest** [reply]
> > @SuisChan I have added the token as a necessary search
> 
> Token is hardcoded, and not hard to find. Each version, even architecture, has its own uniq token

Thanks for the information @SuisChan. Have you already reversed the playplay endpoint ? Do you know where is used the deobfuscated key or how it's done ?

**Reply by SuisChan** [reply]
> Do you know where is used the deobfuscated key or how it's done ?

At the end of the deobfuscation function 😅 

> Have you already reversed the playplay endpoint ?

Working on it.

**Reply by cth-latest** [reply]
> > Do you know where is used the deobfuscated key or how it's done ?
> 
> At the end of the deobfuscation function 😅

Yeah of course aha. But i mean does it used as API Key somewhere for filestorage resolving, or file decrypting, or something else ?

**Reply by SuisChan** [reply]
> or file decrypting

Yes, it's just AES keys, used for content decrypt, usually audio tracks. 

> filestorage

Reversed this part of the app, long story, but irrelevant.

**Reply by cth-latest** [reply]
If i'm not wrong audio files are now hosted on akamaized, downloaded, and decrypted with the licenseKey received in the playplay response ?

**Reply by SuisChan** [reply]
Exactly

**Reply by SuisChan** [reply]
This is weird, because the token is hardcoded, easy to find in the binary. How many tests I ran, it never changed...

**Reply by cth-latest** [reply]
> This is weird, because the token is hardcoded, easy to find in the binary. How many tests I ran, it never changed...

Sorry for that, has you can see i've deleted my message because after checking seems GooglePlay has force updated the mobile application for a minor version changes.

I will do more tests tomorrow to be sûre that was the problem, thanks anyway i will let you updated.

**Reply by SuisChan** [reply]
> I will do more tests tomorrow

My tests show that a token is something like an identifier of a specific encryption state (?) (of a specific version & architecture, and therefore cannot be changed) because the key requested with a different token is not decrypted correctly.

**Reply by cth-latest** [reply]
> > I will do more tests tomorrow
> 
> My tests show that a token is something like an identifier of a specific encryption state (?) (of a specific version & architecture, and therefore cannot be changed) because the key requested with a different token is not decrypted correctly.

Have you found how is decrypted the file song hosted on akamaized ? Do the token is the key or a such salt ?, maybe the file is encrypted and cached on the server side for each different token.

**Reply by SuisChan** [reply]
> Have you found how is decrypted the file song hosted on akamaized ?

The file is encrypted once, and the key is always the same. 

> Do the token is the key or a such salt ?

Looks like it's just an identifier, and you only need to know file_id[:16], maybe this is the salt. 


And I know what cache_id is, I guess

**Reply by cth-latest** [reply]
@SuisChan  If you agree to make a small summary of the information and processes you could find around the playplay endpoint it would be really appreciated, I will update the main thread.

**Reply by SuisChan** [reply]
Okay, but a little later

**Reply by KateYogurt** [reply]
> I don't know the token, but I sent the request with random token anyway, and it returned 502. 
So is setting a token a must or just optional?

![Request 1](https://user-images.githubusercontent.com/102638369/160844541-9e7c0cb5-cf94-49d7-97e2-149fcbd56ba8.png)

**Reply by iscle** [reply]
@KateYogurt check out the playplay branch in my profile, that one has the request working

**Reply by KateYogurt** [reply]
> @KateYogurt check out the playplay branch in my profile, that one has the request working

Thank you! I am checking now. But it seems that the token is fixed?

**Reply by iscle** [reply]
The token has been extracted by reverse engineering a spotify compiled binary. It looks like it is paired with the decryption algorithm, also extracted but not reverse engineered yet. The version code is also paired to the decryption algorithm.

**Comment by sashahilton00** [comment]
@Iscle in the interests of avoiding stepping on any toes, it might be a good idea to replace a chunk of the file ID/decryption keys above with `...` - I think we're probably kosher with documenting the reverse engineering given that it's for educational/interoperability purposes, not sure if the resulting keys are integral to that.

**Reply by iscle** [reply]
I had to delete the comment since the keys still showed up on the edit history. Thanks for the tip!

**Reply by devgianlu** [reply]
> @devgianlu please do not link to my code. I have open legal concerns with Spotify, so doing that is not helping.

Sorry, done.

**Comment by justfoolingaround** [comment]
Any updates on this?

Has this already been done or has it encountered a roadblock?

(Nevermind, I got notified about the DMCA takedown *after* a deobfuscation algorithm was created.)

**Comment by ghost** [comment]
since the above user contacted me privately, I would like to add some context to this comment:

~~~
https://github.com/librespot-org/librespot-java/discussions/421/comments/1812935#discussioncomment-1813151
~~~

no one to my knowledge has reversed the PlayPlay code, at least not publicly. @SuisChan  was close, but they only captured a binary blob rather than implementing or reversing a solution. to anyone reading this page, I would recommend to use the old Shannon code, or to pursue a Widevine solution. Any questions feel free to contact me privately.

**Reply by rugameuser** [reply]
The solution from @SuisChan works fine for decrypting 96kb Vorbis files, but it may not work for other files

**Reply by SuisChan** [reply]
> but it may not work for other files

As far as I remember, it worked quite ok with any quality, something must have changed in recent time (don't know what, haven't touched it for several months).

**Reply by ghost** [reply]
yes, it may "work", but that doesn't change the fact of anything I said in my original comment. its still relying in a binary blob dump, rather than a fully open source solution.

**Reply by SuisChan** [reply]
Well, if you can do it better than others, then go for it.

**Reply by iscle** [reply]
> Also as I already mentioned, just decrypting the Widevine would be a better approach, one I have done several times with other websites.

Can you decrypt widevine with an open-source solution?

**Reply by SuisChan** [reply]
> @SuisChan and I already have done better, by implementing the Shannon code, which is a fully open source solution, granted not using HTTP. Also as I already mentioned, just decrypting the Widevine would be a better approach, one I have done several times with other websites.

The TCP part has not been used in official applications for a long time (with some exceptions) and will probably be removed in the future as unnecessary.  Although everything is working now, it does not mean that it will continue to be so. 

About widevine, I definitely agree with you that this is a good, convenient and easily implemented option,  but with a view as an open source project, this option is not very suitable due to the dependence in a private key, which will be immediately revoked by Google if it gets into the public space.

**Reply by ghost** [reply]
thats why you design an API in such a way that the private key is kept private, as I have done:

https://godocs.io/2a.pages.dev/mech/widevine#New_Module

and provide instruction to user where they can download a private key, or how to dump their own key. or you could even run a Widevine proxy, which some users are already doing, which keeps the private key private as well

**Comment by baltitenger** [comment]
Did anyone get further with this? @SuisChan do you perhaps still have the extracted code / could share what methods and tools you used to find it?

**Comment by jpochyla** [comment]
I was wondering if it'd be possible to use the same approach as the official Web Playback SDK (login through API key, WS communication), but handle the DRM playback through official means, like the FairPlay support in AVPlayer on Apple, some libwidevine on Linux, etc. What do you think?

**Reply by jpochyla** [reply]
Unfortunately WV and FP support only up to 256kbit streams in Spotify, doesn't seem like worth the work.

**Comment by 3052** [comment]
OK according to this:

https://github.com/mIwr/SwiftySpot/blob/94a57db83e7fb4635fa9e64508ee62dcbd4a45be/Example/RhythmRider/Controller/PlaybackController.swift#L499

the `obfuscatedKey` appears to just be used as is. and the decryption seems to be normal AES-128-CTR:

https://github.com/p0rterB/YAD/blob/cbaa733f5bf569822bf62f88ead4f0d6435b7e04/Sources/YAD.swift#L15

has anyone tried this yet? seems too good to be true.

**Reply by SuisChan** [reply]
There's nothing new, just my old code wrapped in the Swift language

**Reply by 3052** [reply]
yeah I see, sorry for the noise

**Reply by SuisChan** [reply]
It's okay, at least you're trying 👌

**Reply by baltitenger** [reply]
@SuisChan by the way, was this extracted from liborbit-jni-spotify.so of the android app?

**Reply by SuisChan** [reply]
Extracted and lifted, but there is a better way than this, maybe I'll write about it someday

**Comment by baltitenger** [comment]
@3052 I think in that code the aes part is just the file decryption, the important bit would be the unwrapKey function, seemingly implemented in c++ (decompiled) which might get taken down just like the previous guy's code.

**Reply by devgianlu** [reply]
@3052 This is correct, the decryption key is returned by the current shannon approach. For playplay they took that key and "obfuscated" it with an hardcoded key present in the client itself. Last guy that published the decompiled code was taken down, if, perhaps, the "obfuscation" scheme turns out to be a standard chiper and I think I'd be less of a problem, but maybe it's the hardcoded key that angers them.

**Reply by justfoolingaround** [reply]
It likely is some sort of standard cipher. The extracted C++ code simply takes the `obfuscated_key` and the `file_id` (as far as I can tell). Additionally there's a table that it uses internally that is supposedly associated with the "token" part sent to Spotify (token is likely an IV that derives that table?). So we're assuming a suite that takes at least 2 inputs other than the obfuscated key.

I think the said hardcoded key isn't quite the problem.  The problem may be some sort of relation between that key and the internal table.

**Reply by baltitenger** [reply]
I think it's more likely that this is an asymmetric encryption scheme and what you're calling the internal table is the private key, the token sent to spotify is the public key. Also, as far as I can tell, the file id is only used near the end of the processData function, so I think it could be split so that the main decryption function only depends on the private key and the encrypted key returned by spotify.

**Comment by DevLARLEY** [comment]
Is this what you have to do decrypt the file?:
+ Get the obfuscated_key from the request.
+ De-obfuscate it using the program by SuisChan
+ Decrypt it with AES-128 CTR, by using the de-obfuscated key as the IV and the file_id as the key

My problem is that the file_id isn't 16 bytes long (I'm guessing the `safeFile.id` is the file_id in bytes)

**Reply by DevLARLEY** [reply]
This is what i have:
```python
    file_id = "bda2281034dec5c42ee97db33588bd08e0b0979f"

    req = playplay_pb2.PlayPlayLicenseRequest(
        version=3,
        token=b"\x01\xe12\xca\xe5'\xbd!b\x0e\x82/XQI2",
        interactivity=playplay_pb2.Interactivity.INTERACTIVE,
        content_type=playplay_pb2.AUDIO_TRACK
    )

    r = requests.post(
        url=f"https://gew4-spclient.spotify.com/playplay/v1/key/{file_id}",
        headers={
            "Authorization": f"Bearer {token_manager.get_access_token()}",
        },
        data=req.SerializeToString()
    )

    resp = playplay_pb2.PlayPlayLicenseResponse()
    resp.ParseFromString(r.content)

    # ================ Decrypter program ====================
    print(file_id, resp.obfuscated_key.hex())
    key = bytes.fromhex("c63d99e107c815e0c3c992ada8868a1e")
    # =======================================================

    # IV from: https://github.com/kokarare1212/librespot-python/blob/3b46fe560ad829b976ce63e85012cff95b1e0bf3/librespot/audio/decrypt.py#L11
    cipher = AES.new(
        key=key,
        mode=AES.MODE_CTR,
        counter=Counter.new(
            128,
            initial_value=int.from_bytes(b'r\xe0g\xfb\xdd\xcb\xcfw\xeb\xe8\xbcd?c\r\x93', "big")
        )
    )
    decrypted_buffer = cipher.decrypt(open(file_id, 'rb').read())
    with open(f'{file_id}.ogg', 'wb') as fw:
        fw.write(decrypted_buffer)
```
Doesn't work though

**Reply by SuisChan** [reply]
> req = playplay_pb2.PlayPlayLicenseRequest(\n        version=3,\n        token=b\"\\x01\\xe12\\xca\\xe5'\\xbd!b\\x0e\\x82/XQI2\",\n        interactivity=playplay_pb2.Interactivity.INTERACTIVE,\n        content_type=playplay_pb2.AUDIO_TRACK\n    )

You must use the provided token to request obfuscated keys (from the deobfuscator info), and the version in the request must be 2, not 3

**Reply by DevLARLEY** [reply]
THANKS! the token is actually the same just in bytes but i didn't know that i had to change the version...

**Reply by DevLARLEY** [reply]
mpv fails at playback, and ffmpeg says:
`[ogg @ 0000021e2509ec00] Codec not found`
but VLC can play it!
EDIT: works on every OGG_VORBIS quality:

The updated and working code:
```python
file_id = "12b4c30164479fd1dd8f41462311420c420470e3"

req = playplay_pb2.PlayPlayLicenseRequest(
    version=2,
    token=bytes.fromhex("01e132cae527bd21620e822f58514932"),
    interactivity=playplay_pb2.Interactivity.INTERACTIVE,
    content_type=playplay_pb2.AUDIO_TRACK
)

r = requests.post(
    url=f"https://gew4-spclient.spotify.com/playplay/v1/key/{file_id}",
    headers={
        "Authorization": f"Bearer {token_manager.get_access_token()}",
    },
    data=req.SerializeToString()
)

resp = playplay_pb2.PlayPlayLicenseResponse()
resp.ParseFromString(r.content)

output = subprocess.check_output([
    'playplay/playplay.exe',
    file_id,
    resp.obfuscated_key.hex()
])
key = bytes.fromhex(output.strip().decode('utf-8'))

cipher = AES.new(
    key=key,
    mode=AES.MODE_CTR,
    counter=Counter.new(
        128,
        initial_value=int.from_bytes(b'r\xe0g\xfb\xdd\xcb\xcfw\xeb\xe8\xbcd?c\r\x93', "big")
    )
)

decrypted_buffer = cipher.decrypt(open(file_id, 'rb').read())
with open(f'{file_id}.ogg', 'wb') as fw:
    fw.write(decrypted_buffer[167:])
```

**Reply by DevLARLEY** [reply]
I'm also trying to download the AAC_24 file, but it doesn't seem playable. A sample can be found here: [https://forum.videohelp.com/threads/415261-Unplayable-AAC-file](https://forum.videohelp.com/threads/415261-Unplayable-AAC-file)
Maybe someone could help me out here? There's NOTHING on the internet regarding this mysterious file header.

**Reply by 3052** [reply]
Contact me on discord

**Reply by DevLARLEY** [reply]
sent

**Reply by DevLARLEY** [reply]
Ok it seems like some tracks (`https://open.spotify.com/track/31qgVdvSqTQ7unwQQngycB` or `https://open.spotify.com/track/2qSkIjg1o9h3YT9RAgYN75`) fail to decrypt. 
Can anyone reproduce this?

**Reply by 3052** [reply]
plays fine in the browser

**Reply by DevLARLEY** [reply]
I'm talking about the OGG_VORBIS version of the song.

**Reply by DevLARLEY** [reply]
Already requested song keys aren't requested again, instead the endpoint `GET https://gew4-spclient.spotify.com/playplay/v1/playable/<file_id>`  is queried without a request body or response.

**Reply by SuisChan** [reply]
> fail to decrypt.\nCan anyone reproduce this?

This is true, unfortunately. The decryptor is not perfect, it can decrypt many keys, but some of them will fail

I have a workaround and it works with all tracks that I've tested so far and never had any problems with decryption

but that's a topic for another time, I don't plan to publish it yet because Spotify will still try to shutdown the repo like last time

Maybe the right time for release will be when lossless appears on the platform? 🤔

**Reply by DevLARLEY** [reply]
@3052 I'm gonna need a copy!

**Reply by justfoolingaround** [reply]
@DevLARLEY Hey, I just wanted to let you know that the reason that `mpv` or `ffmpeg` fails to play is probably due to the extra header Spotify adds. Try skipping 167 bytes from the start! 

See: https://github.com/justfoolingaround/spotivents/blob/master/spotivents/streamer.py#L112

Also good luck on this! And if you do find a solution please notify me as well!

**Reply by DevLARLEY** [reply]
Thanks, @justfoolingaround!
All of my current spotify downloading knowledge can be found here: [https://github.com/DevLARLEY/spotify-oggmp4-dl](https://github.com/DevLARLEY/spotify-oggmp4-dl)

**Reply by staniel359** [reply]
@SuisChan @DevLARLEY Hey.
Do you know where I can get a working version of key deobfuscator (`playplay.exe`)?

**Reply by scorp867** [reply]
> @3052 I'm gonna need a copy!

Are you still active? I need to know if this is still working....if yes then where can I find the playplay.exe from? If no then what are the other ways the deobfuscate the key? Please reply

**Comment by 3052** [comment]
I found another decompilation here:

https://github.com/es3n1n/re-unplayplay/blob/master/src/lib/reversals/decrypt_key.cpp

**Reply by ycq0125** [reply]
> I found another decompilation here:
> 
> https://github.com/es3n1n/re-unplayplay/blob/master/src/lib/reversals/decrypt_key.cpp

Repository unavailable due to DMCA takedown
